net/instaweb/http/redirect_following_url_async_fetcher.cc

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Author: oschaaf@we-amp.com (Otto van der Schaaf)
+
+#include "net/instaweb/http/public/redirect_following_url_async_fetcher.h"
+
+#include "net/instaweb/http/public/async_fetch.h"
+#include "net/instaweb/rewriter/public/rewrite_options.h"
+#include "pagespeed/kernel/base/basictypes.h"
+#include "pagespeed/kernel/base/null_message_handler.h"
+#include "pagespeed/kernel/base/string.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/kernel/http/google_url.h"
+
+// TODO(oschaaf): inlining & intent should  be persisted across redirects.
+namespace net_instaweb {
+
+const int64 RedirectFollowingUrlAsyncFetcher::kUnset =
+  GOOGLE_LONGLONG(0x7FFFFFFFFFFFFFFF);
+
+class RedirectFollowingUrlAsyncFetcher::RedirectFollowingFetch
+    : public SharedAsyncFetch {
+ public:
+  RedirectFollowingFetch(
+      RedirectFollowingUrlAsyncFetcher* redirect_following_fetcher,
+      AsyncFetch* base_fetch, const GoogleString& url,
+      const GoogleString& context_url, MessageHandler* message_handler)
+      : SharedAsyncFetch(base_fetch),
+        redirect_following_fetcher_(redirect_following_fetcher),
+        received_redirect_status_code_(false),
+        urls_seen_(new StringSet()),
+        url_(url),
+        gurl_(url),
+        base_fetch_(base_fetch),
+        context_url_(context_url),
+        message_handler_(message_handler),
+        max_age_(kUnset) {
+    GoogleString sanitized = GoogleUrl::Sanitize(url);
```

Comment:
My goal with `GoogleUrl::Sanitize` was to get the urls in a comparable form for checking that we do not follow redirects we already followed. Reading up on `Spec()`, that seems pretty thorough indeed, and it will save some work here. I'll change this to use `Spec()`.